src/Plugins/Nop.Plugin.Widgets.FacebookPixel/Services/FacebookPixelService.cs

```diff
@@ -801,6 +904,391 @@ public async Task PrepareCompleteRegistrationScriptAsync()
+                var (productPrice, _, _, _) = await _priceCalculationService.GetFinalPriceAsync(product, customer, includeDiscounts: false);
+                var (price, _) = await _taxService.GetProductPriceAsync(product, productPrice);
+                var currentCurrency = await _workContext.GetWorkingCurrencyAsync();
+                var priceValue = await _currencyService.ConvertFromPrimaryStoreCurrencyAsync(price, currentCurrency);
+                var currency = currentCurrency?.CurrencyCode;
+
+                var eventObject = new ConversionsEventCustomData
+                {
+                    ContentCategory = categoryName,
+                    ContentIds = new List<string> { sku },
+                    ContentName = product?.Name,
+                    ContentType = "product",
+                    Contents = new List<object>
+                    {
+                        new
+                        {
+                            id = sku,
+                            quantity = quantity,
+                            item_price = priceValue
+                        }
+                    },
+                    Currency = currency,
+                    Value = priceValue
+                };
+
+                //prepare conversions event api parameter
+                var eventName = item.ShoppingCartTypeId == (int)ShoppingCartType.ShoppingCart
+                    ? FacebookPixelDefaults.ADD_TO_CART
+                    : FacebookPixelDefaults.ADD_TO_WISHLIST;
+
+                var param = new ConversionsEvent
+                {
+                    Data = new List<ConversionsEventDatum>
+                    {
+                        new ConversionsEventDatum
+                        {
+                            EventName = eventName,
+                            EventTime = new DateTimeOffset(item.CreatedOnUtc).ToUnixTimeSeconds(),
+                            EventSourceUrl = _webHelper.GetThisPageUrl(true),
+                            ActionSource = "website",
+                            UserData = await GetUserDataAsync(),
+                            CustomData = eventObject
+                        }
+                    }
+                };
+
+                await SendTrackedEventAsync(eventName, param, item.StoreId);
+
+                return true;
+            });
+        }
+
+        /// <summary>
+        /// Send "Purchase" events
+        /// </summary>
+        /// <param name="order">Order</param>
+        /// <returns>A task that represents the asynchronous operation</returns>
+        public async Task SendPurchaseEventAsync(Order order)
+        {
+            await HandleFunctionAsync(async () =>
+            {
+                //check whether the purchase was initiated by the customer
+                var customer = await _workContext.GetCurrentCustomerAsync();
+                if (order.CustomerId != customer.Id)
+                    return false;
+
+                //prepare event object
```

Comment:
Now preparing of event objects (for all events) is duplicated for the script and for the conversions API, I suggest to combine them, perhaps even using common models for this.